### PR TITLE
[NOD-497] Implement GHOSTDAG

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

**Comment by someone235** [comment]
Please change `blockSet.bluest` to use `*blockNode.less`

File: blockdag/ghostdag.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+				// This is a sanity check that validates that a blue
+				// block's blue anticone is not already larger than K.
+				if candidateBluesAnticoneSizes[block] > dag.dagParams.K {
+					return nil, errors.New("found blue anticone size larger than k")
+				}
+			}
+		}
+
+		if possiblyBlue {
```
Comment (svarogg) [inline blockdag/ghostdag.go]: I don't like the fact that by this line if `possiblyBlue = true`, we know for sure that it is blue, and in general the exit-conditions the above loop has.
Lets discuss alternatives face-to-face/zoom.

**Comment by someone235** [comment]
in `blueBlockWindow` change in the comment phantom to GHOSTDAG

File: blockdag/reachability.go [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+	start := ri.start
+	for i, size := range sizes {
+		intervals[i] = newReachabilityInterval(start, start+size-1)
+		start += size
+	}
+	return intervals, nil
+}
+
+// splitWithExponentialBias splits this interval to len(sizes) parts
+// by some allocation rule. This method expects sum(sizes) to be
```
Comment (svarogg) [inline blockdag/reachability.go]: "some allocation rule"  ->  "the allocation rule described further down this comment" (or something similar)

File: blockdag/reachability.go [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	return ri.splitExact(biasedSizes)
+}
+
+// exponentialFractions returns a fraction of each size in sizes
+// as follows:
+//   fraction[i] = 2^size[i] / sum(2^size[j])
+// In the code below the above equation is divided by 2^max(size)
+// to avoid exploding numbers. Note that in 1 / 2^(max(size)-size[i])
+// we divide 1 by potentially a very large number, which will
+// result in loss of float precision. This is not a problem--all
```
Comment (svarogg) [inline blockdag/reachability.go]: "This  is not a problem--all": chagne `--` to ` - ` (space, dash, space)

File: blockdag/reachability.go [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+
+	// subtreeSize is a helper field used only during reindexing
+	// (expected to be 0 any other time).
+	// See countSubtrees for further details.
+	subtreeSize uint64
+}
+
+func newReachabilityTreeNode(blockNode *blockNode) *reachabilityTreeNode {
+	interval := newReachabilityInterval(1, math.MaxUint64-1)
+	remainingInterval := newReachabilityInterval(interval.start, interval.end-1)
```
Comment (svarogg) [inline blockdag/reachability.go]: Please add a comment explaining why these starting values make sense

File: blockdag/ghostdag_test.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	"github.com/kaspanet/kaspad/dagconfig"
+	"github.com/kaspanet/kaspad/util"
+	"reflect"
+	"sort"
+	"testing"
+)
+
+type testBlockData struct {
+	parents                []string
+	id                     string //id is a virtual entity that is used only for tests so we can define relations between blocks without knowing their hash
```
Comment (stasatdaglabs) [inline blockdag/ghostdag_test.go]: Missing space between // and id.

File: blockdag/ghostdag_test.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+// TestGHOSTDAG iterates over several dag simulations, and checks
+// that the blue score, blue set and selected parent of each
+// block are calculated as expected.
+func TestGHOSTDAG(t *testing.T) {
+	dagParams := dagconfig.SimNetParams
+
+	tests := []struct {
+		k            uint32
+		dagData      []*testBlockData
+		expectedReds []string
```
Comment (stasatdaglabs) [inline blockdag/ghostdag_test.go]: Please put expectedReds above dagData to match the order of definitions below.

File: blockdag/reachabilitystore.go [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+		}
+	}
+
+	return nil
+}
+
+func (store *reachabilityStore) deserializeReachabilityData(
+	serializedReachabilityDataBytes []byte, destination *reachabilityData) error {
+
+	serializedReachabilityData := bytes.NewBuffer(serializedReachabilityDataBytes)
```
Comment (someone235) [inline blockdag/reachabilitystore.go]: calling it `serializedReachabilityData` suggests that it's a byte slice, while it's actually a reader. I would rename it to `r`

File: blockdag/reachabilitystore.go [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+	if !daghash.ZeroHash.IsEqual(parentHash) {
+		parentReachabilityData, ok := store.reachabilityDataByHash(parentHash)
+		if !ok {
+			return errors.Errorf("parent reachability data not found for hash: %s", parentHash)
+		}
+		destination.treeNode.parent = parentReachabilityData.treeNode
+	}
+
+	// Deserialize the amount of children
+	childAmount, err := wire.ReadVarInt(r)
```
Comment (someone235) [inline blockdag/reachabilitystore.go]: In `wire` we usually use the word `count` and not `amount`

File: blockdag/test_utils.go [hunk 9]
```diff
@@ -251,3 +251,59 @@ func LoadBlocks(filename string) (blocks []*util.Block, err error) {
 
 	return
 }
+
+// OpTrueAddress returns an address pointing to a P2SH anyone-can-spend script
+func OpTrueAddress(prefix util.Bech32Prefix) (util.Address, error) {
```
Comment (someone235) [inline blockdag/test_utils.go]: This can be unexported

File: blockdag/test_utils.go [hunk 10]
```diff
@@ -251,3 +251,59 @@ func LoadBlocks(filename string) (blocks []*util.Block, err error) {
...
+	block, err := dag.BlockForMining(blockTransactions)
+	if err != nil {
+		return nil, err
+	}
+	block.Header.Timestamp = dag.NextBlockMinimumTime()
+	return block, nil
+}
+
+// GenerateDeterministicExtraNonceForTest returns a unique deterministic extra nonce for coinbase data, in order to create unique coinbase transactions.
+func GenerateDeterministicExtraNonceForTest() uint64 {
```
Comment (someone235) [inline blockdag/test_utils.go]: This can be unexported

File: blockdag/ghostdag.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	for _, parent := range node.parents {
+		if parent == node.selectedParent {
+			continue
+		}
+		anticoneSet.add(parent)
+		anticoneSlice = append(anticoneSlice, parent)
+		queue = append(queue, parent)
+	}
+	for len(queue) > 0 {
+		var current *blockNode
```
Comment (dereeno) [inline blockdag/ghostdag.go]: Maybe extract the handling of each parent to another function to improve readability.

File: blockdag/reachability.go [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+	i := fb.findIndex(block)
+	if i == 0 {
+		// No candidate to contain block
+		return false
+	}
+
+	candidate := fb[i-1]
+	blockInterval := block.treeNode.interval
+	candidateInterval := candidate.treeNode.interval
+	return candidateInterval.isAncestorOf(blockInterval)
```
[the remark this answers is not in the document]
Comment (dereeno) [reply blockdag/reachability.go]: shouldn't it be the other way around?
`candidate.treeNode.isAncestorOf(block.treeNode)`?
@stasatdaglabs

File: blockdag/reachabilitystore.go [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+	return reachabilityData.futureCoveringSet, nil
+}
+
+func (store *reachabilityStore) reachabilityDataByHash(hash *daghash.Hash) (*reachabilityData, bool) {
+	reachabilityData, ok := store.loaded[*hash]
+	return reachabilityData, ok
+}
+
+// flushToDB writes all dirty reachability data to the database. If all writes
+// succeed, this clears the dirty set.
```
[the remark this answers is not in the document]
Comment (dereeno) [reply blockdag/reachabilitystore.go]: .

File: blockdag/blocknode.go [hunk 14]
```diff
@@ -75,6 +76,9 @@ type blockNode struct {
 	// blueScore is the count of all the blue blocks in this block's past
 	blueScore uint64
 
+	// bluesAnticoneSizes is a map holding the set of blues affected by this block and their modified blue anticone size.
+	bluesAnticoneSizes map[daghash.Hash]uint32
```
[the remark this answers is not in the document]
Comment (dereeno) [reply blockdag/blocknode.go]: .

File: blockdag/ghostdag.go (hunk 1 above)
Comment (dereeno) [reply blockdag/ghostdag.go]: .

File: blockdag/reachability.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply blockdag/reachability.go]: You're right, yes, sorry.

File: blockdag/reachabilitystore.go [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+type reachabilityData struct {
+	treeNode          *reachabilityTreeNode
+	futureCoveringSet futureCoveringBlockSet
+}
+
+type reachabilityStore struct {
+	dag    *BlockDAG
+	dirty  map[daghash.Hash]struct{}
+	loaded map[daghash.Hash]*reachabilityData
+	mtx    sync.RWMutex
```
[the remark this answers is not in the document]
Comment (dereeno) [reply blockdag/reachabilitystore.go]: .

File: blockdag/ghostdag.go [hunk 16]
```diff
@@ -5,75 +5,28 @@ import (
...
-
 // ghostdag runs the GHOSTDAG protocol and updates newNode.blues,
 // newNode.selectedParent and newNode.bluesAnticoneSizes accordingly.
 // The function updates newNode.blues by iterating over the blocks in
 // the anticone of newNode.selectedParent (which is the parent with the
 // highest blue score) and adds any block to newNode.blues if by adding
 // it to newNode.blues these conditions will not be violated:
 //
-// 1) |anticone-of-candidate-block ∩ blueset-of-newNode| ≤ K
+// 1) |anticone-of-candidate-block ∩ blueSet-of-newNode| ≤ K
```
Comment (someone235) [inline blockdag/ghostdag.go]: Please write in comments "blue-set" or "blue set" (depends on the context)

File: blockdag/ghostdag.go [hunk 17]
```diff
@@ -164,3 +117,59 @@ func (dag *BlockDAG) ghostdag(newNode *blockNode) (selectedParentAnticone []*blo
 	newNode.blueScore = newNode.selectedParent.blueScore + uint64(len(newNode.blues))
 	return selectedParentAnticone, nil
 }
+
+// selectedParentAnticone returns the blocks in the anticone of the selected parent of the given node.
+// The function work as follows.
+// We start by adding all parents of the node (other than the selected parent) to a process queue.
+// For each node in the queue:
+//   we check whether it is in the past of the selected parent.
+//   If not, we add the node to the resulting anticone-set and add queue it for processing.
```
Comment (someone235) [inline blockdag/ghostdag.go]: I believe the "add" is redundant in "add queue"

File: blockdag/reachability.go [hunk 18]
```diff
@@ -264,8 +268,8 @@ func (rtn *reachabilityTreeNode) reindexIntervals() ([]*reachabilityTreeNode, er
 		subtreeSize = current.countSubtrees()
 	}
 
-	// Apply the interval down the subtree
-	return current.propagateInterval(current.interval)
+	// Propagate the interval up the subtree
```
Comment (someone235) [inline blockdag/reachability.go]: maybe change "up" to "onto" or "to"? Directions are confusing

File: blockdag/reachability.go [hunk 19]
```diff
@@ -341,14 +345,13 @@ func (rtn *reachabilityTreeNode) countSubtrees() uint64 {
...
+// propagateInterval propagates the new interval using a BFS traversal.
+// Subtree intervals are recursively allocated according to subtree sizes and
+// the allocation rule in splitWithExponentialBias. This method returns
 // a list of reachabilityTreeNodes modified by it.
-func (rtn *reachabilityTreeNode) propagateInterval(
-	interval *reachabilityInterval) ([]*reachabilityTreeNode, error) {
+func (rtn *reachabilityTreeNode) propagateInterval() ([]*reachabilityTreeNode, error) {
 
-	rtn.setInterval(interval)
+	rtn.setInterval(rtn.interval)
```
Comment (someone235) [inline blockdag/reachability.go]: Please explain why do you reset the interval

File: blockdag/blockset.go [hunk 20]
```diff
@@ -142,7 +142,7 @@ func (bs blockSet) bluest() *blockNode {
 	for _, node := range bs {
 		if bluestNode == nil ||
 			node.blueScore > maxScore ||
-			(node.blueScore == maxScore && daghash.Less(node.hash, bluestNode.hash)) {
+			(node.blueScore == maxScore && node.less(bluestNode)) {
```
Comment (someone235) [inline blockdag/blockset.go]: Please remove blue score comparison

**Pull request merged**